### PR TITLE
Add support to multiple keyspaces using Ecto prefixes

### LINKED ISSUE #34: Multiple keyspaces support
Hey! I just discovered Exandra while randomly looking to `xandra` dependents on Hex and it seems it would simplify a lot some codebases we're working on, but that would require support for querying across multiple keyspaces.

Right now, I see it's mainly meant to be used with a single keyspace (since a keyspace is required to be passed in `start_link`), but I already verified that it's possible to query tables on other keyspaces by just using

```
from "<keyspace>.<table>"
```

Would it be possible to map the different keyspaces onto [Ecto prefixes](https://hexdocs.pm/ecto/multi-tenancy-with-query-prefixes.html)? From the documentation it seems that MySQL maps them on different databases, which are more or less the equivalents of keyspaces. From the tests it [seems to be supported](https://github.com/vinniefranco/exandra/blob/main/test/exandra/connection_test.exs#L1017) at least in migrations, but passing `prefix: :foo` in `Repo.all` doesn't seem to work (though it sets the prefix on the returned schema).

If you think this is possible and it would be a desirable feature to have, I can try to tackle this with a PR.

### PR DESCRIPTION
**Pull request by rbino**
This allows using Exandra to query schemas across keyspaces.
Close #34.

### CODE REVIEW

**Comment by rbino** [comment]
I could also potentially drop the requirement for the keyspace in `start_link/1` in this PR, since we can now connect and query multiple keyspaces (it could just be optional).

What do you think @vinniefranco @whatyouhide?

File: lib/exandra.ex [hunk 1]
```diff
@@ -119,6 +119,12 @@ defmodule Exandra do
   Cassandra/Scylla, and allow you to run multiple queries in a single request.
   See `Exandra.Batch` for more information and examples.
 
+  ## Multiple keyspaces using prefixes
+
+  You can use [query prefixes](https://hexdocs.pm/ecto/multi-tenancy-with-query-prefixes.html) to
+  query different keyspaces using the same schemas. Note that, as indicated in the Ecto docs,
+  migrations must be run for _each_ prefix in this case.
```
Comment (whatyouhide) [inline lib/exandra.ex:126]:
```suggestion
  query different keyspaces using the same schemas. As pointed out in the Ecto docs,
  migrations must be run for *each* prefix in this case.
```

File: lib/exandra/connection.ex [hunk 2]
```diff
@@ -822,6 +821,11 @@ defmodule Exandra.Connection do
     "DROP #{quote_name(name)}"
   end
 
+  defp table_name(sources, idx) do
```
Comment (whatyouhide) [inline lib/exandra/connection.ex:824]: We never seem to use `idx`. Why are we not defining `table_name/1` and hard-coding `elem(sources, 0)` on the line below? 🤔

File: test/exandra/connection_test.exs [hunk 3]
```diff
@@ -210,6 +210,21 @@ defmodule Exandra.ConnectionTest do
     end
   end
 
+  test "all with prefix" do
+    query =
+      Schema |> select([r], r.x) |> Queryable.to_query() |> Map.put(:prefix, "prefix") |> plan()
```
Comment (whatyouhide) [inline test/exandra/connection_test.exs:215]: I think this should work?

```suggestion
      Schema |> select([r], r.x) |> Ecto.Query.put_query_prefix("prefix")
```

File: test/exandra/connection_test.exs [hunk 4]
```diff
@@ -210,6 +210,21 @@ defmodule Exandra.ConnectionTest do
...
+
+    assert all(query) == ~s{SELECT x FROM prefix.schema}
+
+    query =
+      Schema |> from(prefix: "first") |> select([r], r.x) |> Map.put(:prefix, "prefix") |> plan()
+
+    assert all(query) == ~s{SELECT x FROM first.schema}
```
Comment (whatyouhide) [inline test/exandra/connection_test.exs:222]: What's the use case of having a test for setting a prefix and then overriding it?

File: test/exandra/migrations_test.exs [hunk 5]
```diff
@@ -127,5 +127,44 @@ defmodule Exandra.MigrationsTest do
 
       assert %{num_rows: 0} = TestRepo.query!(query, [keyspace, "rollbackable"])
     end
+
+    test "with prefix", %{start_opts: start_opts, xandra_conn: xandra_conn} do
```
Comment (whatyouhide) [inline test/exandra/migrations_test.exs:131]: Nice test! 👏

File: test/exandra/prefix_test.exs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  setup do
+    start_link_supervised!({TestRepo, host: "localhost", port: @port, keyspace: "unused"})
+    :ok
+  end
+
+  describe "schema prefix" do
+    setup do
+      # We're only interested in the prepared statement here, so we stub execution
+      XandraMock
+      |> stub(:execute, fn _conn, _stmt, _values, _opts -> {:ok, %Xandra.Void{}} end)
```
Comment (whatyouhide) [inline test/exandra/prefix_test.exs:28]:
```suggestion
      stub(XandraMock :execute, fn _conn, _stmt, _values, _opts -> {:ok, %Xandra.Void{}} end)
```

File: test/exandra/prefix_test.exs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+      :ok
+    end
+
+    test "adds a keyspace to the table" do
+      XandraMock
+      |> expect(:prepare, fn _conn, stmt, _opts ->
```
Comment (whatyouhide) [inline test/exandra/prefix_test.exs:35]:
```suggestion
      expect(XandraMock, :prepare, fn _conn, stmt, _opts ->
```

File: test/exandra/prefix_test.exs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        assert "INSERT INTO foo.my_schema (my_string, id) VALUES (?, ?) " = stmt
+        {:ok, %Xandra.Prepared{}}
+      end)
+
+      TestRepo.insert(%MySchema{my_string: "string"})
```
Comment (whatyouhide) [inline test/exandra/prefix_test.exs:40]:
```suggestion
      assert {:ok, _} = TestRepo.insert(%MySchema{my_string: "string"})
```

File: test/exandra/prefix_test.exs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    end
+
+    test "can be overridden by query options" do
+      XandraMock
+      |> expect(:prepare, fn _conn, stmt, _options ->
+        assert "INSERT INTO bar.my_schema (my_string, id) VALUES (?, ?) " = stmt
+        {:ok, %Xandra.Prepared{}}
+      end)
+
+      TestRepo.insert(%MySchema{my_string: "string"}, prefix: "bar")
```
Comment (whatyouhide) [inline test/exandra/prefix_test.exs:50]:
```suggestion
      assert {:ok, _} = TestRepo.insert(%MySchema{my_string: "string"}, prefix: "bar")
```

File: test/exandra/prefix_test.exs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        assert "INSERT INTO foo.my_schema (my_string, id) VALUES (?, ?) " = stmt
+        {:ok, %Xandra.Prepared{}}
+      end)
+
+      TestRepo.insert(%MySchema{my_string: "string"})
+    end
+
+    test "can be overridden by query options" do
+      XandraMock
+      |> expect(:prepare, fn _conn, stmt, _options ->
```
Comment (whatyouhide) [inline test/exandra/prefix_test.exs:45]:
```suggestion
      expect(XandraMock, :prepare, fn _conn, stmt, _options ->
```

File: test/exandra/querying_test.exs [hunk 11]
```diff
@@ -120,5 +120,12 @@ defmodule Exandra.QueryingTest do
       assert sql == "DELETE FROM my_schema WHERE id = ?"
       assert params == [Ecto.UUID.dump!(uuid)]
     end
+
+    test "with prefix" do
+      query = from(s in MySchema, prefix: "prefix")
+
+      {sql, _params} = TestRepo.to_sql(:all, query)
+      assert sql =~ "FROM prefix.my_schema"
```
Comment (whatyouhide) [inline test/exandra/querying_test.exs:128]: We can do this, right?

```suggestion
      assert String.ends_with?(sql, "FROM prefix.my_schema")
```

**Review by whatyouhide — APPROVED** [review]
Looks really good to me!! I left a few questions, and I’m committing a few style-related suggestions here and there, but we're basically there as far as I'm concerned. 

> I could also potentially drop the requirement for the keyspace in start_link/1 in this PR, since we can now connect and query multiple keyspaces (it could just be optional).

I think it would be better to do this in another PR but yeah, I’d be down to do this.

File: test/exandra/prefix_test.exs [hunk 12]
```diff
@@ -24,30 +24,27 @@ defmodule Exandra.PrefixTest do
   describe "schema prefix" do
     setup do
       # We're only interested in the prepared statement here, so we stub execution
-      XandraMock
-      |> stub(:execute, fn _conn, _stmt, _values, _opts -> {:ok, %Xandra.Void{}} end)
+      stub(XandraMock :execute, fn _conn, _stmt, _values, _opts -> {:ok, %Xandra.Void{}} end)
```
Comment (rbino) [inline test/exandra/prefix_test.exs:27]: Missing `,` here

File: lib/exandra/connection.ex (hunk 2 above)
Comment (rbino) [reply lib/exandra/connection.ex:824]: That's what I started with, I just had the doubt if we could possibly need to pick another table from sources, but from what I understand `sources` can currently be only a 1-tuple given joins and subqueries are not supported. I'm going to drop `idx` and hardcode `0`.

File: test/exandra/connection_test.exs (hunk 4 above)
Comment (rbino) [reply test/exandra/connection_test.exs:222]: This tests that the prefix from `from` is used even if another prefix is set in the query. I took some of these tests from the `ecto_sql` [adapter tests](https://github.com/elixir-ecto/ecto_sql/blob/b1212914a7cb695ce842184080b97348f85c2772/test/ecto/adapters/postgres_test.exs#L1097).

File: test/exandra/querying_test.exs (hunk 11 above)
Comment (rbino) [reply test/exandra/querying_test.exs:128]: Yes, that would be better since it's more strict

**Review by whatyouhide — APPROVED** [review]

**Pull request merged**